### PR TITLE
fix: prevent proposing batches with 1 random request

### PR DESCRIPTION
**Pull request by jorgemmsilva**
there was an edge case where nodes could propose batches with a single request that could not be executed (always skipped). This could deadlock the consensus logic.

also contains the logic from https://github.com/iotaledger/wasp/pull/2656, that should be merged first

### CODE REVIEW

File: packages/vm/core/evm/evmimpl/internal.go [hunk 1]
```diff
@@ -77,6 +77,12 @@ func createEmulator(ctx isc.Sandbox, l2Balance *l2Balance) *emulator.EVMEmulator
 
 // IMPORTANT: Must only be called from the ISC VM (when the request is done executing)
 func AddFailedTx(ctx isc.Sandbox, tx *types.Transaction, receipt *types.Receipt) {
+	if tx == nil {
+		panic("nil tx")
+	}
+	if receipt == nil {
+		panic("nil receipt")
+	}
```
Comment (dessaya) [inline packages/vm/core/evm/evmimpl/internal.go:85]: This left intentionally? Seems like something added just for diagnosing a specific issue.

**Review by dessaya — APPROVED** [review]
Code looks ok, no idea about the logic though :stuck_out_tongue:

File: packages/vm/core/evm/evmimpl/internal.go (hunk 1 above)
Comment (jorgemmsilva) [reply packages/vm/core/evm/evmimpl/internal.go:85]: it was left on purpose, I spend a bunch of time trying to figure out where the "segfault" came from. our "panic handling magic" makes it very hard to follow the stack traces.
basically, this can happen when a  request with a bad nonce enters the VM, EVM panics, but no receipt is produced.

File: packages/chain/mempool/mempool.go [hunk 2]
```diff
@@ -590,20 +590,26 @@ func (mpi *mempoolImpl) handleConsensusProposalForChainHead(recv *reqConsensusPr
...
+		recv.Respond(refs)
 		return
 	}
+
 	//
 	// Wait for any request.
-	mpi.waitReq.WaitAny(recv.ctx, func(req isc.Request) {
-		recv.Respond([]*isc.RequestRef{isc.RequestRefFromRequest(req)})
+	mpi.waitReq.WaitAny(recv.ctx, func(_ isc.Request) {
+		mpi.handleConsensusProposalForChainHead(recv)
```
Comment (kape1395) [inline packages/chain/mempool/mempool.go:612]: Why is this loop needed?
